// @gooddata/sdk-backend-tiger/esm/backend/workspace/permissions/index.js
import { __awaiter, __generator } from "tslib";
var TigerWorkspacePermissionsFactory = /** @class */ (function () {
    function TigerWorkspacePermissionsFactory(authCall, workspace) {
        this.authCall = authCall;
        this.workspace = workspace;
    }
    TigerWorkspacePermissionsFactory.prototype.getPermissionsForCurrentUser = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var response, permissions, _b, canViewWorkspace, canAnalyzeWorkspace, canManageWorkspace, canExportReport;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.authCall(function (client) {
                            return client.entities.getEntityWorkspaces({ id: _this.workspace, metaInclude: ["permissions"] });
                        })];
                    case 1:
                        response = _c.sent();
                        permissions = (_a = response.data.data.meta.permissions) !== null && _a !== void 0 ? _a : [];
                        _b = getPermission(permissions), canViewWorkspace = _b.canViewWorkspace, canAnalyzeWorkspace = _b.canAnalyzeWorkspace, canManageWorkspace = _b.canManageWorkspace, canExportReport = _b.canExportReport;
                        return [2 /*return*/, {
                                //disabled for tiger for now
                                canCreateReport: false,
                                canExportReport: canExportReport,
                                canUploadNonProductionCSV: false,
                                canManageACL: false,
                                canManageDomain: false,
                                canInviteUserToProject: false,
                                canCreateScheduledMail: false,
                                canManageScheduledMail: false,
                                canListUsersInProject: false,
                                //based on group: VIEW
                                canAccessWorkbench: canViewWorkspace,
                                canExecuteRaw: canViewWorkspace,
                                //based on group: ANALYZE
                                canCreateVisualization: canAnalyzeWorkspace,
                                canManageAnalyticalDashboard: canAnalyzeWorkspace,
                                canCreateAnalyticalDashboard: canAnalyzeWorkspace,
                                canManageMetric: canAnalyzeWorkspace,
                                canManageReport: canAnalyzeWorkspace,
                                canRefreshData: canAnalyzeWorkspace,
                                //based on group: MANAGE
                                canManageProject: canManageWorkspace,
                                //NOTE: Data source MANAGE in future
                                canInitData: canManageWorkspace,
                            }];
                }
            });
        });
    };
    return TigerWorkspacePermissionsFactory;
}());
export { TigerWorkspacePermissionsFactory };
function getPermission(permissions) {
    var canViewWorkspace = hasPermission(permissions, "VIEW");
    var canAnalyzeWorkspace = hasPermission(permissions, "ANALYZE");
    var canManageWorkspace = hasPermission(permissions, "MANAGE");
    var canExportReport = hasPermission(permissions, "EXPORT");
    return {
        canViewWorkspace: canViewWorkspace,
        canAnalyzeWorkspace: canAnalyzeWorkspace,
        canManageWorkspace: canManageWorkspace,
        canExportReport: canExportReport,
    };
}
function hasPermission(permissions, need) {
    return permissions.indexOf(need) >= 0;
}
