// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/DraggableCreatePanelItem.js
import { __read } from "tslib";
// (C) 2022 GoodData Corporation
import React from "react";
import cx from "classnames";
import { useDashboardDrag } from "./useDashboardDrag";
/**
 * @internal
 */
export var DraggableCreatePanelItem = function (_a) {
    var Component = _a.Component, disabled = _a.disabled, canDrag = _a.canDrag, onDragStart = _a.onDragStart, onDragEnd = _a.onDragEnd, dragItem = _a.dragItem, hideDefaultPreview = _a.hideDefaultPreview;
    var _b = __read(useDashboardDrag({
        dragItem: dragItem,
        canDrag: canDrag,
        hideDefaultPreview: hideDefaultPreview,
        dragEnd: function (_, monitor) {
            onDragEnd === null || onDragEnd === void 0 ? void 0 : onDragEnd(monitor.didDrop());
        },
        dragStart: onDragStart,
    }, [canDrag, onDragEnd, hideDefaultPreview, dragItem]), 2), isDragging = _b[0].isDragging, dragRef = _b[1];
    return (React.createElement("div", { ref: dragRef, className: cx({ "is-dragging": isDragging }) },
        React.createElement(Component, { disabled: disabled })));
};
