// @gooddata/sdk-backend-tiger/esm/backend/uiFeatures.js
// (C) 2020-2022 GoodData Corporation
/**
 * This is list of feature flags managed on Panther by FeatureHub
 * and keeping their default values for non-managed env
 */
export var TigerFeaturesNames;
(function (TigerFeaturesNames) {
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableSortingByTotalGroup"] = "enableSortingByTotalGroup";
    //string, possible values: disabled, enabledCheckedByDefault, enabledUncheckedByDefault
    TigerFeaturesNames["ADMeasureValueFilterNullAsZeroOption"] = "ADMeasureValueFilterNullAsZeroOption";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableMultipleDates"] = "enableMultipleDates";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableKPIDashboardDeleteFilterButton"] = "enableKPIDashboardDeleteFilterButton";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["DashboardEditModeDevRollout"] = "dashboardEditModeDevRollout";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableTheming"] = "enableTheming";
})(TigerFeaturesNames || (TigerFeaturesNames = {}));
export var DefaultFeatureFlags = {
    enableSortingByTotalGroup: false,
    ADMeasureValueFilterNullAsZeroOption: "EnabledUncheckedByDefault",
    enableMultipleDates: true,
    enableKPIDashboardDeleteFilterButton: false,
    // disable edit mode in gdc-dashboards during development
    dashboardEditModeDevRollout: false,
    enableTheming: true,
};
export var FeatureFlagsValues = {
    enableSortingByTotalGroup: [true, false],
    ADMeasureValueFilterNullAsZeroOption: [
        "Disabled",
        "EnabledCheckedByDefault",
        "EnabledUncheckedByDefault",
    ],
    enableMultipleDates: [true, false],
    enableKPIDashboardDeleteFilterButton: [true, false],
    dashboardEditModeDevRollout: [true, false],
    enableTheming: [true, false],
};
