// @gooddata/sdk-ui-dashboard/esm/widgets/placeholders/types.js
// (C) 2021-2022 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import { newCustomWidget } from "../../model/types/layoutTypes";
/**
 * Tests whether an object is a {@link KpiPlaceholderWidget}.
 *
 * @param obj - object to test
 * @alpha
 */
export function isKpiPlaceholderWidget(obj) {
    return !isEmpty(obj) && obj.customType === "gd-kpi-placeholder";
}
/**
 * @internal
 */
export var KPI_PLACEHOLDER_WIDGET_ID = "__kpiPlaceholder__";
/**
 * @alpha
 */
export function newKpiPlaceholderWidget() {
    return newCustomWidget(KPI_PLACEHOLDER_WIDGET_ID, "gd-kpi-placeholder");
}
/**
 * Tests whether an object is a {@link InsightPlaceholderWidget}.
 *
 * @param obj - object to test
 * @alpha
 */
export function isInsightPlaceholderWidget(obj) {
    return !isEmpty(obj) && obj.customType === "gd-insight-placeholder";
}
/**
 * @internal
 */
export var INSIGHT_PLACEHOLDER_WIDGET_ID = "__insightPlaceholder__";
/**
 * @alpha
 */
export function newInsightPlaceholderWidget() {
    return newCustomWidget(INSIGHT_PLACEHOLDER_WIDGET_ID, "gd-insight-placeholder");
}
/**
 * Tests whether an object is a {@link PlaceholderWidget}.
 *
 * @param obj - object to test
 * @alpha
 */
export function isPlaceholderWidget(obj) {
    return !isEmpty(obj) && obj.customType === "gd-widget-placeholder";
}
/**
 * Tests whether an object is a {@link PlaceholderWidget} and is initial.
 *
 * @param obj - object to test
 * @internal
 */
export function isInitialPlaceholderWidget(obj) {
    return isPlaceholderWidget(obj) && !!obj.isInitial;
}
/**
 * @internal
 */
export var PLACEHOLDER_WIDGET_ID = "__placeholder__";
/**
 * @alpha
 */
export function newPlaceholderWidget() {
    return newCustomWidget(PLACEHOLDER_WIDGET_ID, "gd-widget-placeholder");
}
/**
 * @internal
 */
export function newInitialPlaceholderWidget() {
    return newCustomWidget(PLACEHOLDER_WIDGET_ID, "gd-widget-placeholder", {
        isInitial: true,
    });
}
/**
 * Tests whether an object is any type of placeholder widgets.
 *
 * @param obj - object to test
 * @alpha
 */
export function isAnyPlaceholderWidget(obj) {
    return isPlaceholderWidget(obj) || isInsightPlaceholderWidget(obj) || isKpiPlaceholderWidget(obj);
}
