// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/draggableWidget/useAddInitialSectionHandler.js
// (C) 2022 GoodData Corporation
import { useCallback } from "react";
import { useDashboardSelector, useDashboardDispatch, selectIsLayoutEmpty, addLayoutSection, selectSettings, } from "../../../model";
import { isInsightDraggableListItem, isInsightPlaceholderDraggableItem, isKpiPlaceholderDraggableItem, } from "../types";
import { newInitialPlaceholderWidget } from "../../../widgets";
import { getInsightPlaceholderSizeInfo, getSizeInfo } from "../../../_staging/layout/sizing";
export function useAddInitialSectionHandler() {
    var dispatch = useDashboardDispatch();
    var settings = useDashboardSelector(selectSettings);
    var isLayoutEmpty = useDashboardSelector(selectIsLayoutEmpty);
    return useCallback(function (item) {
        if (isLayoutEmpty) {
            var sizeInfo = void 0;
            if (isInsightDraggableListItem(item)) {
                sizeInfo = getSizeInfo(settings, "insight", item.insight);
            }
            if (isInsightPlaceholderDraggableItem(item)) {
                sizeInfo = getInsightPlaceholderSizeInfo(settings);
            }
            if (isKpiPlaceholderDraggableItem(item)) {
                sizeInfo = getSizeInfo(settings, "kpi");
            }
            if (sizeInfo) {
                dispatch(addLayoutSection(0, {}, [
                    {
                        type: "IDashboardLayoutItem",
                        size: {
                            xl: {
                                gridHeight: sizeInfo.height.default,
                                gridWidth: sizeInfo.width.default,
                            },
                        },
                        widget: newInitialPlaceholderWidget(),
                    },
                ]));
            }
        }
    }, [dispatch, isLayoutEmpty, settings]);
}
