// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/draggableWidget/useNewSectionInsightListItemDropHandler.js
// (C) 2022 GoodData Corporation
import { useCallback } from "react";
import { insightRef, insightTitle } from "@gooddata/sdk-model";
import { getSizeInfo } from "../../../_staging/layout/sizing";
import { selectSettings, useDashboardDispatch, useDashboardSelector, useDashboardCommandProcessing, uiActions, enableInsightWidgetDateFilter, addLayoutSection, } from "../../../model";
export function useNewSectionInsightListItemDropHandler(sectionIndex) {
    var dispatch = useDashboardDispatch();
    var settings = useDashboardSelector(selectSettings);
    var preselectDateDataset = useDashboardCommandProcessing({
        commandCreator: enableInsightWidgetDateFilter,
        errorEvent: "GDC.DASH/EVT.COMMAND.FAILED",
        successEvent: "GDC.DASH/EVT.INSIGHT_WIDGET.FILTER_SETTINGS_CHANGED",
        onSuccess: function (event) {
            dispatch(uiActions.setWidgetLoadingAdditionalDataStopped(event.payload.ref));
        },
        onError: function (event) {
            dispatch(uiActions.setWidgetLoadingAdditionalDataStopped(event.payload.command.payload.ref));
        },
    }).run;
    var addNewSectionWithInsight = useDashboardCommandProcessing({
        commandCreator: addLayoutSection,
        errorEvent: "GDC.DASH/EVT.COMMAND.FAILED",
        successEvent: "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        onSuccess: function (event) {
            var ref = event.payload.section.items[0].widget.ref;
            dispatch(uiActions.selectWidget(ref));
            dispatch(uiActions.setConfigurationPanelOpened(true));
            dispatch(uiActions.setWidgetLoadingAdditionalDataStarted(ref));
            preselectDateDataset(ref, "default");
        },
    }).run;
    return useCallback(function (insight) {
        var sizeInfo = getSizeInfo(settings, "insight", insight);
        addNewSectionWithInsight(sectionIndex, {}, [
            {
                type: "IDashboardLayoutItem",
                widget: {
                    type: "insight",
                    insight: insightRef(insight),
                    ignoreDashboardFilters: [],
                    drills: [],
                    title: insightTitle(insight),
                    description: "",
                    configuration: { hideTitle: false },
                    properties: {},
                },
                size: {
                    xl: {
                        gridHeight: sizeInfo.height.default,
                        gridWidth: sizeInfo.width.default,
                    },
                },
            },
        ]);
    }, [addNewSectionWithInsight, sectionIndex, settings]);
}
