// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/draggableWidget/useNewSectionKpiPlaceholderDropHandler.js
// (C) 2022 GoodData Corporation
import { useCallback } from "react";
import { idRef } from "@gooddata/sdk-model";
import { getSizeInfo } from "../../../_staging/layout/sizing";
import { selectSettings, useDashboardDispatch, useDashboardSelector, useDashboardCommandProcessing, uiActions, addLayoutSection, } from "../../../model";
import { KPI_PLACEHOLDER_WIDGET_ID, newKpiPlaceholderWidget } from "../../../widgets";
export function useNewSectionKpiPlaceholderDropHandler(sectionIndex) {
    var dispatch = useDashboardDispatch();
    var settings = useDashboardSelector(selectSettings);
    var addNewSectionWithKpiPlaceholder = useDashboardCommandProcessing({
        commandCreator: addLayoutSection,
        errorEvent: "GDC.DASH/EVT.COMMAND.FAILED",
        successEvent: "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        onSuccess: function (event) {
            var ref = event.payload.section.items[0].widget.ref;
            dispatch(uiActions.selectWidget(idRef(KPI_PLACEHOLDER_WIDGET_ID)));
            dispatch(uiActions.setConfigurationPanelOpened(true));
            dispatch(uiActions.setKpiDateDatasetAutoOpen(true));
            dispatch(uiActions.setWidgetLoadingAdditionalDataStarted(ref));
        },
    }).run;
    return useCallback(function () {
        var sizeInfo = getSizeInfo(settings, "kpi");
        addNewSectionWithKpiPlaceholder(sectionIndex, {}, [
            {
                type: "IDashboardLayoutItem",
                size: {
                    xl: {
                        gridHeight: sizeInfo.height.default,
                        gridWidth: sizeInfo.width.default,
                    },
                },
                widget: newKpiPlaceholderWidget(),
            },
        ]);
    }, [addNewSectionWithKpiPlaceholder, sectionIndex, settings]);
}
