// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/draggableWidget/useNewSectionInsightPlaceholderDropHandler.js
// (C) 2022 GoodData Corporation
import { useCallback } from "react";
import { idRef } from "@gooddata/sdk-model";
import { getInsightPlaceholderSizeInfo } from "../../../_staging/layout/sizing";
import { selectSettings, useDashboardDispatch, useDashboardSelector, useDashboardCommandProcessing, uiActions, addLayoutSection, } from "../../../model";
import { INSIGHT_PLACEHOLDER_WIDGET_ID, newInsightPlaceholderWidget } from "../../../widgets";
export function useNewSectionInsightPlaceholderDropHandler(sectionIndex) {
    var dispatch = useDashboardDispatch();
    var settings = useDashboardSelector(selectSettings);
    var addNewSectionWithInsightPlaceholder = useDashboardCommandProcessing({
        commandCreator: addLayoutSection,
        errorEvent: "GDC.DASH/EVT.COMMAND.FAILED",
        successEvent: "GDC.DASH/EVT.FLUID_LAYOUT.SECTION_ADDED",
        onSuccess: function () {
            dispatch(uiActions.selectWidget(idRef(INSIGHT_PLACEHOLDER_WIDGET_ID)));
            dispatch(uiActions.setConfigurationPanelOpened(true));
        },
    }).run;
    return useCallback(function () {
        var sizeInfo = getInsightPlaceholderSizeInfo(settings);
        addNewSectionWithInsightPlaceholder(sectionIndex, {}, [
            {
                type: "IDashboardLayoutItem",
                size: {
                    xl: {
                        gridHeight: sizeInfo.height.default,
                        gridWidth: sizeInfo.width.default,
                    },
                },
                widget: newInsightPlaceholderWidget(),
            },
        ]);
    }, [addNewSectionWithInsightPlaceholder, sectionIndex, settings]);
}
