// @gooddata/sdk-ui-dashboard/esm/model/store/config/configSelectors.js
// (C) 2021-2022 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import invariant from "ts-invariant";
var selectSelf = createSelector(function (state) { return state; }, function (state) { return state.config; });
/**
 * Returns dashboard's config.
 *
 * @remarks
 * It is expected that the selector is called only after the config state
 * is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export var selectConfig = createSelector(selectSelf, function (configState) {
    invariant(configState.config, "attempting to access uninitialized config state");
    return configState.config;
});
/**
 * Returns workspace-level configuration for the of the date filter options and presets.
 *
 * @remarks
 * Note: this configuration SHOULD be further augmented by the dashboard-level overrides to obtain
 * the effective date filter configuration.
 *
 * @public
 */
export var selectDateFilterConfig = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.dateFilterConfig) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns settings that are in effect for the current dashboard.
 *
 * @public
 */
export var selectSettings = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.settings) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns locale to use for internationalization of the dashboard.
 *
 * @public
 */
export var selectLocale = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.locale) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns number separators to use when rendering numeric values on charts or KPIs.
 *
 * @public
 */
export var selectSeparators = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.separators) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns the color palette for dashboard charts.
 *
 * @public
 */
export var selectColorPalette = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.colorPalette) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns the object availability configuration for this dashboard.
 *
 * @remarks
 * Only objects that match the availability criteria can appear in selections where user has pick
 * an object to use for some purpose (for instance metric for KPI or date dataset to filter by).
 *
 * @public
 */
export var selectObjectAvailabilityConfig = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.objectAvailability) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns Mapbox token.
 *
 * @internal
 */
export var selectMapboxToken = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.mapboxToken) !== null && _a !== void 0 ? _a : undefined;
});
/**
 * Returns whether the Dashboard is executed in read-only mode.
 *
 * @remarks
 * Read-only mode disables any interactions that can alter the backend data.
 *
 * @public
 */
export var selectIsReadOnly = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.isReadOnly) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether the Dashboard is executed in embedded context.
 *
 * @remarks
 * In embedded mode, some interactions may be disabled.
 *
 * @public
 */
export var selectIsEmbedded = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.isEmbedded) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether the Dashboard is rendered in the export mode.
 * In export mode, some components can be hidden, or rendered differently.
 *
 * @internal
 */
export var selectIsExport = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.isExport) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether the Dashboard is white labeled.
 *
 * @internal
 */
export var selectIsWhiteLabeled = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.isWhiteLabeled) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether the default drills configured on the widgets or implicit drills (eg. drill down) are disabled.
 * This option does not affect drilling enabled by drillableItems.
 *
 * @public
 */
export var selectDisableDefaultDrills = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.disableDefaultDrills) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether filter values in drill events should be resolved.
 *
 * @public
 */
export var selectEnableFilterValuesResolutionInDrillEvents = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.enableFilterValuesResolutionInDrillEvents) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether the save as new button is hidden.
 *
 * @internal
 */
export var selectIsSaveAsNewButtonHidden = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.hideSaveAsNewButton) !== null && _a !== void 0 ? _a : false;
});
//
// FEATURE FLAGS
//
/**
 * Returns date format.
 *
 * @public
 */
export var selectDateFormat = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.responsiveUiDateFormat) !== null && _b !== void 0 ? _b : undefined;
});
/**
 * Returns whether the current user can schedule emails.
 *
 * @public
 */
export var selectEnableKPIDashboardSchedule = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardSchedule) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the current user can share scheduled email to other recipients.
 *
 * @public
 */
export var selectEnableKPIDashboardScheduleRecipients = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardScheduleRecipients) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns current platform edition.
 *
 * @public
 */
export var selectPlatformEdition = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.platformEdition) !== null && _b !== void 0 ? _b : "enterprise";
});
/**
 * Returns whether company logo should be visible in embedded dashboard.
 *
 * @public
 */
export var selectEnableCompanyLogoInEmbeddedUI = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableCompanyLogoInEmbeddedUI) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the export to pdf is enabled.
 *
 * @public
 */
export var selectEnableKPIDashboardExportPDF = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardExportPDF) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the drill to dashboard is enabled.
 *
 * @public
 */
export var selectEnableKPIDashboardDrillToDashboard = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDrillToDashboard) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the save as new dashboard functionality is enabled.
 *
 * @public
 */
export var selectEnableKPIDashboardSaveAsNew = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardSaveAsNew) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether implicit drill to attributes url enabled
 *
 * @public
 */
export var selectEnableClickableAttributeURL = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableClickableAttributeURL) !== null && _b !== void 0 ? _b : true;
});
/**
 * Returns whether drill to url is enabled
 *
 * @public
 */
export var selectEnableKPIDashboardDrillToURL = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDrillToURL) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether drill to insight is enabled
 *
 * @public
 */
export var selectEnableKPIDashboardDrillToInsight = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDrillToInsight) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether implicit drill to attributes url enabled
 *
 * @public
 */
export var selectEnableKPIDashboardImplicitDrillDown = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardImplicitDrillDown) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether Kpi drills in embedded mode are disabled.
 *
 * @public
 */
export var selectHideKpiDrillInEmbedded = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.hideKpiDrillInEmbedded) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether insight export scheduling is enabled.
 *
 * @public
 */
export var selectEnableInsightExportScheduling = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableInsightExportScheduling) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether dashboard edit mode is enabled.
 *
 * @internal
 */
export var selectDashboardEditModeDevRollout = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = !!((_a = state.settings) === null || _a === void 0 ? void 0 : _a.dashboardEditModeDevRollout)) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether analytical dashboard permissions are enabled
 *
 * @internal
 */
export var selectEnableAnalyticalDashboardPermissions = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableAnalyticalDashboardPermissions) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether custom widget heights are enabled
 *
 * @internal
 */
export var selectEnableWidgetCustomHeight = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKDWidgetCustomHeight) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether we should call workspaces workspaces (true) or projects (false).
 *
 * @internal
 */
export var selectEnableRenamingProjectToWorkspace = createSelector(selectConfig, function (state) {
    var _a, _b;
    return !!((_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableRenamingProjectToWorkspace) !== null && _b !== void 0 ? _b : true);
});
/**
 * Returns whether we should call measures metrics (true) or measures (false).
 *
 * @internal
 */
export var selectEnableRenamingMeasureToMetric = createSelector(selectConfig, function (state) {
    var _a, _b;
    return !!((_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableRenamingMeasureToMetric) !== null && _b !== void 0 ? _b : false);
});
/**
 * Returns whether we should hide the pixel perfect experience references.
 *
 * @internal
 */
export var selectShouldHidePixelPerfectExperience = createSelector(selectConfig, function (state) {
    var _a, _b, _c, _d;
    var isHidden = (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.hidePixelPerfectExperience) !== null && _b !== void 0 ? _b : true;
    var isEnabled = (_d = (_c = state.settings) === null || _c === void 0 ? void 0 : _c.enablePixelPerfectExperience) !== null && _d !== void 0 ? _d : false;
    return !isHidden && isEnabled;
});
/**
 * Returns whether we should disable the underline in KPIs when they are drillable.
 *
 * @internal
 */
export var selectDisableKpiDashboardHeadlineUnderline = createSelector(selectConfig, function (state) {
    var _a, _b;
    return !!((_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.disableKpiDashboardHeadlineUnderline) !== null && _b !== void 0 ? _b : false);
});
/**
 * Returns whether unfinished features are allowed.
 *
 * @internal
 */
export var selectAllowUnfinishedFeatures = createSelector(selectConfig, function (state) { return state.allowUnfinishedFeatures || false; });
/**
 * Returns whether analytical designer is enabled.
 *
 * @internal
 */
export var selectIsAnalyticalDesignerEnabled = createSelector(selectConfig, function (state) { var _a; return !!(((_a = state.settings) === null || _a === void 0 ? void 0 : _a.analyticalDesigner) || false); });
/**
 * Returns whether delete button in dashboard attribute filters is visible.
 *
 * @internal
 */
export var selectIsDeleteFilterButtonEnabled = createSelector(selectConfig, function (state) { var _a; return !!(((_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDeleteFilterButton) || false); });
