// @gooddata/sdk-ui-dashboard/esm/presentation/widget/kpi/EditModeDashboardKpi/useOptimisticMeasureUpdate.js
import { __read } from "tslib";
// (C) 2022 GoodData Corporation
import { useEffect, useState } from "react";
import { areObjRefsEqual } from "@gooddata/sdk-model";
import { useDashboardSelector, useDashboardEventsContext, isDashboardCommandStarted, isDashboardKpiWidgetMeasureChanged, isDashboardCommandFailed, selectAllCatalogMeasuresMap, } from "../../../../model";
export function useOptimisticMeasureUpdate(kpiWidget) {
    // ref of the measure the KPI is being updated to
    var _a = __read(useState(), 2), updatingMeasureRef = _a[0], setUpdatingMeasureRef = _a[1];
    var measures = useDashboardSelector(selectAllCatalogMeasuresMap);
    var currentMeasure = measures.get(kpiWidget.kpi.metric);
    var updatingMeasure = updatingMeasureRef && measures.get(updatingMeasureRef);
    var _b = useDashboardEventsContext(), registerHandler = _b.registerHandler, unregisterHandler = _b.unregisterHandler;
    useEffect(function () {
        var onMeasureChangingStarted = {
            eval: function (evt) {
                return (isDashboardCommandStarted(evt) &&
                    evt.payload.command.type === "GDC.DASH/CMD.KPI_WIDGET.CHANGE_MEASURE" &&
                    areObjRefsEqual(evt.payload.command.payload.ref, kpiWidget.ref));
            },
            handler: function (e) {
                setUpdatingMeasureRef(e.payload.command.payload.measureRef);
            },
        };
        var onMeasureChangingEnded = {
            eval: function (evt) {
                return ((isDashboardKpiWidgetMeasureChanged(evt) &&
                    areObjRefsEqual(evt.payload.ref, kpiWidget.ref)) ||
                    (isDashboardCommandFailed(evt) &&
                        evt.payload.command.type === "GDC.DASH/CMD.KPI_WIDGET.CHANGE_MEASURE" &&
                        areObjRefsEqual(evt.payload.command.payload.ref, kpiWidget.ref)));
            },
            handler: function () {
                setUpdatingMeasureRef(undefined);
            },
        };
        registerHandler(onMeasureChangingStarted);
        registerHandler(onMeasureChangingEnded);
        return function () {
            unregisterHandler(onMeasureChangingStarted);
            unregisterHandler(onMeasureChangingEnded);
        };
    }, [kpiWidget.ref, registerHandler, unregisterHandler]);
    var isChangingMeasure = !!updatingMeasureRef;
    // if the current title of the KPI is the same as the title of the current measure, after the update is done,
    // it will take on the title of the new measure, otherwise the title is untouched
    var titleToShow = isChangingMeasure && kpiWidget.title === (currentMeasure === null || currentMeasure === void 0 ? void 0 : currentMeasure.measure.title) && updatingMeasure
        ? updatingMeasure.measure.title
        : kpiWidget.title;
    return {
        isChangingMeasure: isChangingMeasure,
        titleToShow: titleToShow,
    };
}
