// @gooddata/sdk-ui-dashboard/esm/presentation/localization/IntlWrapper.js
// (C) 2007-2022 GoodData Corporation
import React, { useCallback } from "react";
import { IntlProvider } from "react-intl";
import { DefaultLocale, TranslationsCustomizationProvider, } from "@gooddata/sdk-ui";
import { translations } from "./translations";
var formats = {
    time: {
        hhmm: {
            hour: "numeric",
            minute: "2-digit",
        },
    },
    date: {
        shortWithoutYear: {
            day: "numeric",
            month: "short",
        },
        shortWithYear: {
            day: "numeric",
            month: "short",
            year: "numeric",
        },
    },
};
/**
 * @internal
 */
export var IntlWrapper = function (_a) {
    var children = _a.children, _b = _a.locale, locale = _b === void 0 ? DefaultLocale : _b;
    var render = useCallback(function (modifiedTranslations) { return (React.createElement(IntlProvider, { locale: locale, messages: modifiedTranslations, formats: formats }, children)); }, [locale, children]);
    return React.createElement(TranslationsCustomizationProvider, { translations: translations[locale], render: render });
};
