// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/draggableAttributeFilter/DraggableAttributeFilter.js
import { __read } from "tslib";
// (C) 2022 GoodData Corporation
import React, { useMemo } from "react";
import classNames from "classnames";
import { useDashboardSelector, selectIsInEditMode, selectSupportsElementUris } from "../../../model";
import { AttributeFilterDropZoneHint } from "./AttributeFilterDropZoneHint";
import { useDashboardDrag } from "../useDashboardDrag";
import { convertDashboardAttributeFilterElementsUrisToValues } from "../../../_staging/dashboard/legacyFilterConvertors";
export function DraggableAttributeFilter(_a) {
    var FilterComponent = _a.FilterComponent, filter = _a.filter, filterIndex = _a.filterIndex, onAttributeFilterChanged = _a.onAttributeFilterChanged, onAttributeFilterAdded = _a.onAttributeFilterAdded;
    var isInEditMode = useDashboardSelector(selectIsInEditMode);
    var _b = __read(useDashboardDrag({
        dragItem: {
            type: "attributeFilter",
            filter: filter,
            filterIndex: filterIndex,
        },
        canDrag: isInEditMode,
    }, [filter, filterIndex, isInEditMode]), 2), isDragging = _b[0].isDragging, dragRef = _b[1];
    var showDropZones = isInEditMode && !isDragging;
    var supportElementUris = useDashboardSelector(selectSupportsElementUris);
    var filterToUse = useMemo(function () {
        if (supportElementUris) {
            return filter;
        }
        return convertDashboardAttributeFilterElementsUrisToValues(filter);
    }, [filter, supportElementUris]);
    return (React.createElement("div", { className: "draggable-attribute-filter" },
        showDropZones && (React.createElement(AttributeFilterDropZoneHint, { hintPosition: "prev", targetIndex: filterIndex, onAddAttributePlaceholder: onAttributeFilterAdded })),
        React.createElement("div", { className: classNames("dash-filters-notdate", "dash-filters-attribute", {
                "is-dragging": isDragging,
            }), ref: dragRef },
            React.createElement(FilterComponent, { filter: filterToUse, onFilterChanged: onAttributeFilterChanged, isDraggable: isInEditMode })),
        showDropZones && (React.createElement(AttributeFilterDropZoneHint, { hintPosition: "next", targetIndex: filterIndex, onAddAttributePlaceholder: onAttributeFilterAdded }))));
}
