// @gooddata/sdk-ui-dashboard/esm/model/store/backendCapabilities/backendCapabilitiesSelectors.js
// (C) 2021-2022 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import invariant from "ts-invariant";
var selectSelf = createSelector(function (state) { return state; }, function (state) { return state.backendCapabilities; });
/**
 * This selector returns capabilities of the backend with which the dashboard works.
 *
 * @public
 */
export var selectBackendCapabilities = createSelector(selectSelf, function (state) {
    invariant(state.backendCapabilities, "attempting to access uninitialized backend capabilities");
    return state.backendCapabilities;
});
/**
 * This selector returns capability if parent child filtering is enabled.
 *
 * @public
 */
export var selectSupportsElementsQueryParentFiltering = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.supportsElementsQueryParentFiltering) !== null && _a !== void 0 ? _a : false; });
/**
 * Selector for {@link @gooddata/sdk-backend-spi#IBackendCapabilities.supportsKpiWidget}
 *
 * @internal
 */
export var selectSupportsKpiWidgetCapability = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.supportsKpiWidget) !== null && _a !== void 0 ? _a : false; });
/**
 * Selector for {@link @gooddata/sdk-backend-spi#IBackendCapabilities.supportsAccessControl}
 *
 * @internal
 */
export var selectSupportsAccessControlCapability = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.supportsAccessControl) !== null && _a !== void 0 ? _a : false; });
/**
 * Selector for {@link @gooddata/sdk-backend-spi#IBackendCapabilities.supportsHierarchicalWorkspaces}
 *
 * @internal
 */
export var selectSupportsHierarchicalWorkspacesCapability = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.supportsHierarchicalWorkspaces) !== null && _a !== void 0 ? _a : false; });
/**
 * Selector for {@link @gooddata/sdk-backend-spi#IBackendCapabilities.supportsElementUris}
 *
 * @internal
 */
export var selectSupportsElementUris = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.supportsElementUris) !== null && _a !== void 0 ? _a : false; });
/**
 * Selector for {@link @gooddata/sdk-backend-spi#IBackendCapabilities.canExportCsv}
 *
 * @internal
 */
export var selectSupportsExportToCsv = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.canExportCsv) !== null && _a !== void 0 ? _a : false; });
/**
 * Selector for {@link @gooddata/sdk-backend-spi#IBackendCapabilities.canExportXlsx}
 *
 * @internal
 */
export var selectSupportsExportToXlsx = createSelector(selectBackendCapabilities, function (capabilities) { var _a; return (_a = capabilities.canExportXlsx) !== null && _a !== void 0 ? _a : false; });
