// @gooddata/sdk-ui-dashboard/esm/presentation/filterBar/filterBar/hooks/useFilterBarState.js
// (C) 2021-2022 GoodData Corporation
import { __read } from "tslib";
import { useState, useEffect } from "react";
import { CalculatedRowsDefault } from "./useRowsCalculator";
import { useDashboardDispatch, useDashboardSelector, selectFilterBarExpanded, uiActions, } from "../../../../model";
//NOTE: This 1px is size of border bottom on filter bar
var BorderWidth = 1;
//NOTE: This must be same value as $transition-length
var TransitionLength = 200;
export function useFilterBarState() {
    var _a = __read(useState(CalculatedRowsDefault), 2), calculatedRows = _a[0], setCalculatedRows = _a[1];
    var expandedHeight = calculatedRows.expandedHeight, collapsedHeight = calculatedRows.collapsedHeight, rows = calculatedRows.rows;
    var dispatch = useDashboardDispatch();
    var isFilterBarExpanded = useDashboardSelector(selectFilterBarExpanded);
    var scrollable = useScrollable(isFilterBarExpanded);
    var setFilterBarExpanded = function (isExpanded) {
        return dispatch(uiActions.setFilterBarExpanded(isExpanded));
    };
    return {
        rows: rows,
        scrollable: scrollable,
        height: isFilterBarExpanded ? expandedHeight + BorderWidth : collapsedHeight,
        isFilterBarExpanded: isFilterBarExpanded,
        setCalculatedRows: setCalculatedRows,
        setFilterBarExpanded: setFilterBarExpanded,
    };
}
function useScrollable(expanded) {
    var _a = __read(useState(false), 2), scrollable = _a[0], setScrollable = _a[1];
    useEffect(function () {
        var timer = null;
        if (expanded) {
            timer = window.setTimeout(function () { return setScrollable(true); }, TransitionLength);
        }
        else {
            setScrollable(false);
        }
        return function () {
            if (timer !== null) {
                window.clearTimeout(timer);
            }
        };
    }, [expanded]);
    return scrollable;
}
