// @gooddata/sdk-backend-tiger/esm/convertors/fromBackend/afm/DimensionHeaderConverter.js
// (C) 2022 GoodData Corporation
import { isAttributeDescriptor, isMeasureGroupDescriptor, } from "@gooddata/sdk-model";
import { isResultAttributeHeader, isResultMeasureHeader, isResultTotalHeader, JsonApiAttributeOutAttributesGranularityEnum, } from "@gooddata/api-client-tiger";
import { createDateValueFormatter } from "../dateFormatting/dateValueFormatter";
import { toSdkGranularity } from "../dateGranularityConversions";
var supportedSuffixes = Object.keys(JsonApiAttributeOutAttributesGranularityEnum)
    .filter(function (item) { return isNaN(Number(item)); })
    .map(function (key) {
    return JsonApiAttributeOutAttributesGranularityEnum[key];
});
function getGranularity(header) {
    if (!isAttributeDescriptor(header)) {
        return undefined;
    }
    var identifier = header.attributeHeader.formOf.identifier;
    var suffix = identifier.substr(identifier.lastIndexOf(".") + 1);
    return supportedSuffixes.includes(suffix)
        ? toSdkGranularity(suffix)
        : undefined; // not a date attribute
}
function getMeasuresFromDimensions(dimensions) {
    for (var _i = 0, dimensions_1 = dimensions; _i < dimensions_1.length; _i++) {
        var dim = dimensions_1[_i];
        var measureGroup = dim.headers.find(isMeasureGroupDescriptor);
        if (measureGroup) {
            return measureGroup.measureGroupHeader.items;
        }
    }
    return [];
}
export function getTransformDimensionHeaders(dimensions, dateFormatter) {
    var measureDescriptors = getMeasuresFromDimensions(dimensions);
    var dateValueFormatter = createDateValueFormatter(dateFormatter);
    return function (dimensionHeaders) {
        return dimensionHeaders.map(function (dimensionHeader, dimensionIndex) {
            return dimensionHeader.headerGroups.map(function (headerGroup, headerGroupIndex) {
                var granularity = getGranularity(dimensions[dimensionIndex].headers[headerGroupIndex]);
                return headerGroup.headers.map(function (header) {
                    if (isResultAttributeHeader(header)) {
                        return attributeMeasureItem(header, granularity, dateValueFormatter);
                    }
                    if (isResultMeasureHeader(header)) {
                        return measureHeaderItem(header, measureDescriptors);
                    }
                    if (isResultTotalHeader(header)) {
                        return totalHeaderItem(header);
                    }
                    // This code should never be reachable
                    throw new Error("Unexpected type of ResultHeader: " + header);
                });
            });
        });
    };
}
function attributeMeasureItem(header, granularity, dateValueFormatter) {
    return {
        attributeHeaderItem: {
            uri: header.attributeHeader.primaryLabelValue,
            name: granularity
                ? dateValueFormatter(header.attributeHeader.labelValue, granularity)
                : header.attributeHeader.labelValue,
        },
    };
}
function measureHeaderItem(header, measureDescriptors) {
    var _a;
    /*
     * Funny stuff #1 - Tiger sends just the measure index in the measure headers. This is the index of the
     * measure descriptor within the measure group. The code looks up the measure descriptor so that
     * it can then fill in the `name` to the one in the descriptor
     */
    var measureIndex = header.measureHeader.measureIndex;
    return {
        measureHeaderItem: {
            name: (_a = measureDescriptors[measureIndex]) === null || _a === void 0 ? void 0 : _a.measureHeaderItem.name,
            order: measureIndex,
        },
    };
}
function totalHeaderItem(header) {
    return {
        totalHeaderItem: {
            type: header.totalHeader.function,
            name: header.totalHeader.function.toLowerCase(),
        },
    };
}
