// @gooddata/sdk-ui-dashboard/esm/plugins/customizationApis/filtersCustomizer.js
import { DefaultAttributeFiltersCustomizer } from "./attributeFiltersCustomizer";
import { DefaultDateFiltersCustomizer } from "./dateFiltersCustomizer";
var DefaultFiltersCustomizer = /** @class */ (function () {
    function DefaultFiltersCustomizer(logger) {
        var _this = this;
        this.logger = logger;
        this.attributeFiltersCustomizer = new DefaultAttributeFiltersCustomizer(this.logger);
        this.dateFiltersCustomizer = new DefaultDateFiltersCustomizer(this.logger);
        this.sealCustomizer = function () {
            _this.attributeFiltersCustomizer.sealCustomizer();
            _this.dateFiltersCustomizer.sealCustomizer();
            return _this;
        };
    }
    DefaultFiltersCustomizer.prototype.attribute = function () {
        return this.attributeFiltersCustomizer;
    };
    DefaultFiltersCustomizer.prototype.date = function () {
        return this.dateFiltersCustomizer;
    };
    return DefaultFiltersCustomizer;
}());
export { DefaultFiltersCustomizer };
