// @gooddata/sdk-ui-dashboard/esm/model/store/ui/uiReducers.js
import { areObjRefsEqual } from "@gooddata/sdk-model";
var openScheduleEmailDialog = function (state) {
    state.scheduleEmailDialog.open = true;
};
var closeScheduleEmailDialog = function (state) {
    state.scheduleEmailDialog.open = false;
};
var setScheduleEmailDialogDefaultAttachment = function (state, action) {
    state.scheduleEmailDialog.defaultAttachmentRef = action.payload;
};
var resetScheduleEmailDialogDefaultAttachment = function (state) {
    state.scheduleEmailDialog.defaultAttachmentRef = undefined;
};
var openScheduleEmailManagementDialog = function (state) {
    state.scheduleEmailManagementDialog.open = true;
};
var closeScheduleEmailManagementDialog = function (state) {
    state.scheduleEmailManagementDialog.open = false;
};
var openSaveAsDialog = function (state) {
    state.saveAsDialog.open = true;
};
var closeSaveAsDialog = function (state) {
    state.saveAsDialog.open = false;
};
var openShareDialog = function (state) {
    state.shareDialog.open = true;
};
var closeShareDialog = function (state) {
    state.shareDialog.open = false;
};
var openDeleteDialog = function (state) {
    state.deleteDialog.open = true;
};
var closeDeleteDialog = function (state) {
    state.deleteDialog.open = false;
};
var openKpiDeleteDialog = function (state, action) {
    state.kpiDeleteDialog.widgetCoordinates = action.payload;
};
var closeKpiDeleteDialog = function (state) {
    state.kpiDeleteDialog.widgetCoordinates = undefined;
};
var setFilterBarExpanded = function (state, action) {
    state.filterBar.expanded = action.payload;
};
var openKpiAlertDialog = function (state, action) {
    state.kpiAlerts.openedWidgetRef = action.payload;
};
var closeKpiAlertDialog = function (state) {
    state.kpiAlerts.openedWidgetRef = undefined;
};
var openCancelEditModeDialog = function (state) {
    state.cancelEditModeDialog.open = true;
};
var closeCancelEditModeDialog = function (state) {
    state.cancelEditModeDialog.open = false;
};
var highlightKpiAlert = function (state, action) {
    state.kpiAlerts.highlightedWidgetRef = action.payload;
};
var setMenuButtonItemsVisibility = function (state, action) {
    state.menuButton.itemsVisibility = action.payload;
};
var selectWidget = function (state, action) {
    state.selectedWidgetRef = action.payload;
};
var clearWidgetSelection = function (state) {
    state.selectedWidgetRef = undefined;
};
var setConfigurationPanelOpened = function (state, action) {
    state.configurationPanelOpened = action.payload;
};
var setKpiDateDatasetAutoOpen = function (state, action) {
    state.kpiDateDatasetAutoOpen = action.payload;
};
var requestInsightListUpdate = function (state) {
    state.insightListLastUpdateRequested = +new Date();
};
var setWidgetLoadingAdditionalDataStarted = function (state, action) {
    state.widgetsLoadingAdditionalData.push(action.payload);
};
var setWidgetLoadingAdditionalDataStopped = function (state, action) {
    state.widgetsLoadingAdditionalData = state.widgetsLoadingAdditionalData.filter(function (item) { return !areObjRefsEqual(item, action.payload); });
};
var setFilterAttributeSelectionOpen = function (state, action) {
    state.filterAttributeSelectionOpen = action.payload;
};
var selectFilterIndex = function (state, action) {
    state.selectedFilterIndex = action.payload;
};
var clearFilterIndexSelection = function (state) {
    state.selectedFilterIndex = undefined;
};
var setIsDraggingWidget = function (state, action) {
    state.isDraggingWidget = action.payload;
};
var setActiveSectionIndex = function (state, action) {
    state.activeSectionIndex = action.payload;
};
var clearActiveSectionIndex = function (state) {
    state.activeSectionIndex = undefined;
};
export var uiReducers = {
    openScheduleEmailDialog: openScheduleEmailDialog,
    closeScheduleEmailDialog: closeScheduleEmailDialog,
    setScheduleEmailDialogDefaultAttachment: setScheduleEmailDialogDefaultAttachment,
    resetScheduleEmailDialogDefaultAttachment: resetScheduleEmailDialogDefaultAttachment,
    openScheduleEmailManagementDialog: openScheduleEmailManagementDialog,
    closeScheduleEmailManagementDialog: closeScheduleEmailManagementDialog,
    openSaveAsDialog: openSaveAsDialog,
    closeSaveAsDialog: closeSaveAsDialog,
    setFilterBarExpanded: setFilterBarExpanded,
    closeKpiAlertDialog: closeKpiAlertDialog,
    openKpiAlertDialog: openKpiAlertDialog,
    highlightKpiAlert: highlightKpiAlert,
    openShareDialog: openShareDialog,
    closeShareDialog: closeShareDialog,
    openDeleteDialog: openDeleteDialog,
    closeDeleteDialog: closeDeleteDialog,
    openKpiDeleteDialog: openKpiDeleteDialog,
    closeKpiDeleteDialog: closeKpiDeleteDialog,
    setMenuButtonItemsVisibility: setMenuButtonItemsVisibility,
    selectWidget: selectWidget,
    clearWidgetSelection: clearWidgetSelection,
    setConfigurationPanelOpened: setConfigurationPanelOpened,
    setKpiDateDatasetAutoOpen: setKpiDateDatasetAutoOpen,
    requestInsightListUpdate: requestInsightListUpdate,
    setWidgetLoadingAdditionalDataStarted: setWidgetLoadingAdditionalDataStarted,
    setWidgetLoadingAdditionalDataStopped: setWidgetLoadingAdditionalDataStopped,
    setFilterAttributeSelectionOpen: setFilterAttributeSelectionOpen,
    selectFilterIndex: selectFilterIndex,
    clearFilterIndexSelection: clearFilterIndexSelection,
    setIsDraggingWidget: setIsDraggingWidget,
    setActiveSectionIndex: setActiveSectionIndex,
    clearActiveSectionIndex: clearActiveSectionIndex,
    openCancelEditModeDialog: openCancelEditModeDialog,
    closeCancelEditModeDialog: closeCancelEditModeDialog,
};
