// @gooddata/sdk-ui-dashboard/esm/model/commandHandlers/widgets/setDrillForKpiWidgetHandler.js
// (C) 2021-2022 GoodData Corporation
import { __generator } from "tslib";
import { call, put, select } from "redux-saga/effects";
import { kpiWidgetDrillSet } from "../../events/kpi";
import { selectWidgetsMap } from "../../store/layout/layoutSelectors";
import { validateExistingKpiWidget } from "./validation/widgetValidations";
import { layoutActions } from "../../store/layout";
import { validateKpiDrill } from "./validation/kpiDrillValidation";
export function setDrillForKpiWidgetHandler(ctx, cmd) {
    var correlationId, _a, legacyDashboardTabIdentifier, legacyDashboardRef, widgets, kpiWidget, widgetRef, drill;
    return __generator(this, function (_b) {
        switch (_b.label) {
            case 0:
                correlationId = cmd.correlationId;
                _a = cmd.payload, legacyDashboardTabIdentifier = _a.legacyDashboardTabIdentifier, legacyDashboardRef = _a.legacyDashboardRef;
                return [4 /*yield*/, select(selectWidgetsMap)];
            case 1:
                widgets = _b.sent();
                kpiWidget = validateExistingKpiWidget(widgets, cmd, ctx);
                widgetRef = kpiWidget.ref;
                drill = {
                    tab: legacyDashboardTabIdentifier,
                    target: legacyDashboardRef,
                    origin: {
                        type: "drillFromMeasure",
                        measure: kpiWidget.kpi.metric,
                    },
                    type: "drillToLegacyDashboard",
                    transition: "in-place",
                };
                return [4 /*yield*/, call(validateKpiDrill, drill, ctx, cmd)];
            case 2:
                _b.sent();
                return [4 /*yield*/, put(layoutActions.replaceKpiWidgetDrill({
                        ref: widgetRef,
                        drill: drill,
                        undo: {
                            cmd: cmd,
                        },
                    }))];
            case 3:
                _b.sent();
                return [2 /*return*/, kpiWidgetDrillSet(ctx, widgetRef, drill, correlationId)];
        }
    });
}
