// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/Resize/ResizeOverlay.js
import { __assign } from "tslib";
// (C) 2019-2022 GoodData Corporation
import React from "react";
import cx from "classnames";
import { FormattedMessage, defineMessages } from "react-intl";
var messages = defineMessages({
    minHeight: { id: "layout.widget.height.min" },
    maxHeight: { id: "layout.widget.height.max" },
    minWidth: { id: "layout.widget.width.min" },
    maxWidth: { id: "layout.widget.width.max" },
});
var ResizeOverlayStatus;
(function (ResizeOverlayStatus) {
    ResizeOverlayStatus[ResizeOverlayStatus["None"] = 0] = "None";
    ResizeOverlayStatus[ResizeOverlayStatus["Grey"] = 1] = "Grey";
    ResizeOverlayStatus[ResizeOverlayStatus["Active"] = 2] = "Active";
    ResizeOverlayStatus[ResizeOverlayStatus["Error"] = 3] = "Error";
})(ResizeOverlayStatus || (ResizeOverlayStatus = {}));
function getMessage(_a) {
    var reachedHeightLimit = _a.reachedHeightLimit, reachedWidthLimit = _a.reachedWidthLimit;
    if (reachedHeightLimit === "min") {
        return messages.minHeight;
    }
    if (reachedHeightLimit === "max") {
        return messages.maxHeight;
    }
    if (reachedWidthLimit === "min") {
        return messages.minWidth;
    }
    if (reachedWidthLimit === "max") {
        return messages.maxWidth;
    }
}
function getStatus(_a) {
    var isResizingColumnOrRow = _a.isResizingColumnOrRow, isActive = _a.isActive, reachedWidthLimit = _a.reachedWidthLimit, reachedHeightLimit = _a.reachedHeightLimit;
    var status = ResizeOverlayStatus.None;
    if (isResizingColumnOrRow) {
        status = ResizeOverlayStatus.Grey;
        if (isActive) {
            status = ResizeOverlayStatus.Active;
            if (reachedWidthLimit !== "none" || reachedHeightLimit !== "none") {
                status = ResizeOverlayStatus.Error;
            }
        }
    }
    return status;
}
export function ResizeOverlay(props) {
    var status = getStatus(props);
    if (status === ResizeOverlayStatus.None) {
        return null;
    }
    var isInError = status === ResizeOverlayStatus.Error;
    var isActive = status === ResizeOverlayStatus.Active;
    var classes = cx("gd-resize-overlay", {
        active: isActive,
        error: isInError,
    });
    var message = getMessage({
        reachedHeightLimit: props.reachedHeightLimit,
        reachedWidthLimit: props.reachedWidthLimit,
    });
    var errorText = (React.createElement("div", { className: "gd-resize-overlay-text" },
        React.createElement(FormattedMessage, __assign({}, message))));
    return (React.createElement(React.Fragment, null,
        React.createElement("div", { role: "resize-overlay", className: classes }),
        isInError ? errorText : null));
}
