// @gooddata/sdk-ui-dashboard/esm/presentation/dragAndDrop/Resize/useScrollCorrection.js
// (C) 2022 GoodData Corporation
import { __read } from "tslib";
import { useEffect, useMemo, useState } from "react";
var emptyPosition = {
    x: 0,
    y: 0,
};
export function useScrollCorrection(getDimensions, isActive) {
    if (isActive === void 0) { isActive = true; }
    var _a = __read(useState(emptyPosition), 2), currentPosition = _a[0], setCurrentPosition = _a[1];
    var initialPosition = useMemo(function () {
        if (isActive) {
            return getDimensions();
        }
        else {
            return emptyPosition;
        }
    }, [isActive, getDimensions]);
    useEffect(function () {
        var updatePosition = function () {
            setCurrentPosition(getDimensions);
        };
        if (isActive) {
            document.addEventListener("scroll", updatePosition, true);
            updatePosition();
        }
        return function () { return document.removeEventListener("scroll", updatePosition, true); };
    }, [isActive, getDimensions]);
    var scrollCorrection = useMemo(function () {
        return {
            x: currentPosition.x - initialPosition.x,
            y: currentPosition.y - initialPosition.y,
        };
    }, [currentPosition, initialPosition]);
    return {
        initialPosition: initialPosition,
        currentPosition: currentPosition,
        scrollCorrection: scrollCorrection,
    };
}
