// @gooddata/sdk-ui-dashboard/esm/presentation/drill/types.js
import isEmpty from "lodash/isEmpty";
import { isDrillToAttributeUrl, isDrillToCustomUrl, } from "@gooddata/sdk-model";
export function isDrillToUrl(drillDefinition) {
    return isDrillToCustomUrl(drillDefinition) || isDrillToAttributeUrl(drillDefinition);
}
export var DRILL_TARGET_TYPE;
(function (DRILL_TARGET_TYPE) {
    DRILL_TARGET_TYPE["DRILL_TO_DASHBOARD"] = "DrillToDashboard";
    DRILL_TARGET_TYPE["DRILL_TO_INSIGHT"] = "DrillToInsight";
    DRILL_TARGET_TYPE["DRILL_TO_URL"] = "DrillToUrl";
})(DRILL_TARGET_TYPE || (DRILL_TARGET_TYPE = {}));
export function isDrillToDashboardConfig(item) {
    return !isEmpty(item) && item.dashboard !== undefined;
}
export function isDrillToInsightConfig(item) {
    return !isEmpty(item) && item.insightRef !== undefined;
}
export function isDrillToCustomUrlConfig(item) {
    return !isEmpty(item) && item.customUrl !== undefined;
}
export function isDrillToAttributeUrlConfig(item) {
    return !isEmpty(item) && item.insightAttributeDisplayForm !== undefined;
}
export function isDrillToUrlConfig(item) {
    return !isEmpty(item) && item.urlDrillTarget !== undefined;
}
export var AttributeDisplayFormType;
(function (AttributeDisplayFormType) {
    AttributeDisplayFormType["HYPERLINK"] = "GDC.link";
    AttributeDisplayFormType["GEO_PUSHPIN"] = "GDC.geo.pin";
})(AttributeDisplayFormType || (AttributeDisplayFormType = {}));
export function isAvailableDrillTargetMeasure(obj) {
    return !isEmpty(obj) && obj.measure !== undefined;
}
