// @gooddata/sdk-ui-dashboard/esm/presentation/filterBar/attributeFilter/dashboardDropdownBody/configuration/hooks/useConnectingAttributes.js
// (C) 2022 GoodData Corporation
import { useEffect, useMemo } from "react";
import { queryConnectingAttributes, useDashboardQueryProcessing, } from "../../../../../../model";
/**
 * @internal
 */
export function useConnectingAttributes(currentFilterDisplayForm, neighborFiltersDisplayForms) {
    var pairs = useMemo(function () {
        return neighborFiltersDisplayForms.map(function (neighborDisplayForm) { return [
            currentFilterDisplayForm,
            neighborDisplayForm,
        ]; });
    }, [neighborFiltersDisplayForms, currentFilterDisplayForm]);
    var _a = useDashboardQueryProcessing({
        queryCreator: queryConnectingAttributes,
    }), getConnectingAttributes = _a.run, connectingAttributes = _a.result, connectingAttributesStatus = _a.status, connectingAttributesError = _a.error;
    useEffect(function () {
        getConnectingAttributes(pairs);
    }, [pairs, getConnectingAttributes]);
    var connectingAttributesLoading = useMemo(function () {
        return connectingAttributesStatus === "pending" || connectingAttributesStatus === "running";
    }, [connectingAttributesStatus]);
    return {
        connectingAttributes: connectingAttributes,
        connectingAttributesLoading: connectingAttributesLoading,
        connectingAttributesError: connectingAttributesError,
    };
}
