// @gooddata/sdk-backend-tiger/esm/backend/organization/settings.js
import { __assign, __awaiter, __generator } from "tslib";
// (C) 2022 GoodData Corporation
import { isUnexpectedError } from "@gooddata/sdk-backend-spi";
import { convertApiError } from "../../utils/errorHandling";
import { unwrapSettingContent } from "../../convertors/fromBackend/SettingsConverter";
var OrganizationSettingsService = /** @class */ (function () {
    function OrganizationSettingsService(authCall) {
        this.authCall = authCall;
    }
    OrganizationSettingsService.prototype.setWhiteLabeling = function (whiteLabeling) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.setSetting("whiteLabeling", whiteLabeling)];
            });
        });
    };
    OrganizationSettingsService.prototype.setLocale = function (locale) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.setSetting("locale", { value: locale })];
            });
        });
    };
    OrganizationSettingsService.prototype.getSettings = function () {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                            return [2 /*return*/, client.entities.getAllEntitiesOrganizationSettings({})];
                        }); }); })];
                    case 1:
                        data = (_a.sent()).data;
                        return [2 /*return*/, data.data.reduce(function (result, setting) {
                                var _a;
                                var _b;
                                return __assign(__assign({}, result), (_a = {}, _a[setting.id] = unwrapSettingContent((_b = setting.attributes) === null || _b === void 0 ? void 0 : _b.content), _a));
                            }, {})];
                }
            });
        });
    };
    OrganizationSettingsService.prototype.setSetting = function (id, content) {
        return __awaiter(this, void 0, void 0, function () {
            var error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 6]);
                        return [4 /*yield*/, this.getSetting(id)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.updateSetting(id, content)];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 6];
                    case 3:
                        error_1 = _a.sent();
                        if (!isUnexpectedError(error_1)) return [3 /*break*/, 5];
                        // if such settings is not defined
                        return [4 /*yield*/, this.createSetting(id, content)];
                    case 4:
                        // if such settings is not defined
                        _a.sent();
                        return [2 /*return*/];
                    case 5: throw convertApiError(error_1);
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    OrganizationSettingsService.prototype.getSetting = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) {
                        return client.entities.getEntityOrganizationSettings({
                            id: id,
                        });
                    })];
            });
        });
    };
    OrganizationSettingsService.prototype.updateSetting = function (id, content) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) {
                        return client.entities.updateEntityOrganizationSettings({
                            id: id,
                            jsonApiOrganizationSettingInDocument: {
                                data: {
                                    type: "organizationSetting",
                                    id: id,
                                    attributes: {
                                        content: content,
                                    },
                                },
                            },
                        });
                    })];
            });
        });
    };
    OrganizationSettingsService.prototype.createSetting = function (id, content) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) {
                        return client.entities.createEntityOrganizationSettings({
                            jsonApiOrganizationSettingInDocument: {
                                data: {
                                    type: "organizationSetting",
                                    id: id,
                                    attributes: {
                                        content: content,
                                    },
                                },
                            },
                        });
                    })];
            });
        });
    };
    return OrganizationSettingsService;
}());
export { OrganizationSettingsService };
